### PR TITLE
bottomless2: perform store job

### PR DESCRIPTION
**Pull request by MarinPostma**
- implement scheduler
- move bottomless2 to libsql-wal
- add async fs methods
- implement perform job

### CODE REVIEW

File: .cargo/config.toml
Comment (MarinPostma) [inline .cargo/config.toml:1]: just quieting warnings

File: libsql-wal/Cargo.toml [hunk 1]
```diff
@@ -19,3 +19,12 @@ thiserror = "1.0.58"
 tracing = "0.1.40"
 walkdir = "2.5.0"
 zerocopy = { version = "0.7.32", features = ["derive", "alloc"] }
+tokio = { version = "1", features = ["full"] }
```
Comment (LucioFranco) [inline libsql-wal/Cargo.toml:22]: do we need all the features?

File: libsql-wal/Cargo.toml [hunk 2]
```diff
@@ -19,3 +19,12 @@ thiserror = "1.0.58"
...
+tempfile = "3.10.1"
```
Comment (LucioFranco) [inline libsql-wal/Cargo.toml:23]: dev dep?

File: libsql-wal/src/bottomless/mod.rs [hunk 3]
```diff
@@ -71,7 +81,7 @@ impl<S: Storage + 'static> BottomlessLoop<S> {
                         Err(e) => {
                             // job panicked. report and exit process. The program is crippled, from
                             // now on, so we just exit, and hope to restart on a fresh state.
-                            tracing::error!("fatal error: bottomless job panicked: {e}");
+                            tracing::error!("fatal error: bottomles job panicked: {e}");
```
Comment (LucioFranco) [inline libsql-wal/src/bottomless/mod.rs:84]:
```suggestion
                            tracing::error!("fatal error: bottomless job panicked: {e}");
```

File: libsql-wal/src/io/file.rs [hunk 4]
```diff
@@ -0,0 +1,246 @@
...
+        offset: u64,
+    ) -> impl Future<Output = (B, Result<()>)> + Send;
+}
+
+impl FileExt for File {
+    fn write_all_at(&self, buf: &[u8], offset: u64) -> Result<()> {
+        let mut written = 0;
+
+        while written != buf.len() {
+            written += nix::sys::uio::pwrite(self, &buf[written..], offset as _)?;
```
Comment (LucioFranco) [inline libsql-wal/src/io/file.rs:47]: dont you need to write at offset + written?

File: libsql-wal/src/io/mod.rs [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+        create_new: bool,
+        read: bool,
+        write: bool,
+        path: &Path,
+    ) -> io::Result<Self::File>;
+
+    // todo: create an async counterpart
+    fn tempfile(&self) -> io::Result<Self::TempFile>;
+    fn now(&self) -> DateTime<Utc>;
+    fn uuid(&self) -> Uuid;
```
Comment (LucioFranco) [inline libsql-wal/src/io/mod.rs:29]: uuid is part of the io system 🤣

Comment (MarinPostma) [reply libsql-wal/src/io/mod.rs:29]: it's randomness

Comment (MarinPostma) [reply libsql-wal/src/io/mod.rs:29]: randomness is an input

Comment (MarinPostma) [reply libsql-wal/src/io/mod.rs:29]: actually I'll probably just expose a range 🤔

File: libsql-wal/src/io/file.rs (hunk 4 above)
Comment (MarinPostma) [reply libsql-wal/src/io/file.rs:47]: hmm yeah, I need to write a test that catches that

Comment (MarinPostma) [reply libsql-wal/src/io/file.rs:47]: I was unable to write a test to verify that, no matter how big the payload, it's writing everything in one go...

File: libsql-wal/Cargo.toml (hunk 2 above)
Comment (MarinPostma) [reply libsql-wal/Cargo.toml:23]: nope, it's being used

File: libsql-wal/Cargo.toml (hunk 1 above)
Comment (MarinPostma) [reply libsql-wal/Cargo.toml:22]: Probably not, but we're compiling them anyway. I'll trim later.

File: libsql-wal/src/bottomless/job.rs [hunk 6]
```diff
@@ -0,0 +1,442 @@
...
+    let mut pages = segment.index().stream();
+    // todo: use Frame::Zeroed somehow, so that header is aligned?
+    let mut buffer = Box::new(ZeroCopyBuf::<Frame>::new_uninit());
+    let mut out_index = fst::MapBuilder::memory();
+    let mut current_offset = 0;
+
+    while let Some((page_no_bytes, offset)) = pages.next() {
+        let page_no = u32::from_be_bytes(page_no_bytes.try_into().unwrap());
+        let (b, ret) = segment.read_frame_offset_async(offset as _, buffer).await;
+        ret.unwrap();
```
Comment (LucioFranco) [inline libsql-wal/src/bottomless/job.rs:149]: whats this unwrap for?

Comment (MarinPostma) [reply libsql-wal/src/bottomless/job.rs:149]: Uh yeah, I should return an io error here, I'll do in a follow up

File: libsql-wal/src/bottomless/job.rs [hunk 7]
```diff
@@ -0,0 +1,442 @@
...
+            fn tempfile(&self) -> std::io::Result<Self::TempFile> {
+                StdIO(()).tempfile()
+            }
+
+            fn now(&self) -> DateTime<Utc> {
+                DateTime::UNIX_EPOCH
+            }
+
+            fn uuid(&self) -> Uuid {
+                dbg!();
```
Comment (LucioFranco) [inline libsql-wal/src/bottomless/job.rs:328]:
```suggestion
```

File: libsql-wal/src/wal.rs [hunk 8]
```diff
@@ -43,7 +52,7 @@ impl<FS: FileSystem> WalManager for LibsqlWalManager<FS> {
             .registry
             .clone()
             .open(db_path.as_ref())
-            .map_err(Into::into)?;
+            .map_err(|e| dbg!(e.into()))?;
```
Comment (LucioFranco) [inline libsql-wal/src/wal.rs:55]: can remove this?

**Review by LucioFranco — APPROVED** [review]

**Pull request merged**